introduce JUnit tag "flaky", to exclude tests from running.

soloturn opened the pull request:
ExampleTest.testClientCreation fails without real reason, mark it flaky at the moment. permit this tag in engine-tests/build.gadle.kts to separate test runs for normal integration tests, and ones which are not stable on github, for currently unknown reasons. normally of course it would be better to fix these tests - but accepting pull requests is coupled on tests running through. now two commands instead of one run all integration tests:

    gradle integrationTest
    gradle integrationTestFlaky

please do not lightheartedly mark tests as flaky.

soloturn committed:
introduce JUnit tag "flaky", to exclude tests from running.

ExampleTest.testClientCreation fails without real reason, mark it flaky
at the moment. permit this tag in engine-tests/build.gadle.kts to
separate test runs for normal integration tests, and ones which are
not stable on github, for currently unknown reasons. normally of course
it would be better to fix these tests - but accepting pull requests
is coupled on tests running through. now two commands instead of one
run all integration tests:

    gradle integrationTest
    gradle integrationTestFlaky

please do not lightheartedly mark tests as flaky.

jdrueckert committed:
ci: split integration test stage into non-flaky and flaky

jdrueckert commented on engine-tests/build.gradle.kts:
```diff
@@ -2,6 +2,13 @@
 
 // Engine tests are split out due to otherwise quirky project dependency issues with module tests extending engine tests
+// while locally all tests should be run, when building via github, tests can be run separated in the github pipeline
+// file. for integrationtests a tag "flaky" was instroduced to mark tests which are not always successful on github.
```
```suggestion
// file. for integrationtests a tag "flaky" was introduced to mark tests which frequently fail pipelines
```

jdrueckert commented on engine-tests/build.gradle.kts:
```diff
@@ -105,18 +112,30 @@ tasks.register<Test>("unitTest") {
     dependsOn(tasks.getByPath(":extractNatives"))
     group = "Verification"
-    description = "Runs integration tests (slow) tagged with 'MteTest' or 'TteTest'"
+    description = "Runs integration tests (slow) tagged with 'MteTest' or 'TteTest', no tests tagged 'flaky'."
```
```suggestion
    description = "Runs integration tests (slow) tagged with 'MteTest' or 'TteTest', exclude tests tagged 'flaky'."
```

jdrueckert commented on engine-tests/build.gradle.kts:
```diff
@@ -105,18 +112,30 @@ tasks.register<Test>("unitTest") {
+tasks.register<Test>("integrationTestFlaky") {
+    dependsOn(tasks.getByPath(":extractNatives"))
+    group = "Verification"
+    description = "Runs flaky integration tests tagged with 'MteTest', 'TteTest' AND tag 'flaky'."
```
```suggestion
    description = "Runs integration tests tagged with 'flaky' and either 'MteTest' or 'TteTest'."
```

jdrueckert reviewed (commented):
In general, this looks good to me.
I made a few suggestions based on my personal preferences of wording and expression, they are not blocking though.

I split the CI stage using the new gradle task you introduced which works nicely :+1: 
As you can see the tests are executed by the stage matching their tagging:
![image](https://github.com/MovingBlocks/Terasology/assets/29981695/ad0ec4ae-29b1-4737-8eef-f30edbecae8a)

[`org.terasology.engine.integrationenvironment.ExampleTest.testClientConnection()`](https://jenkins.terasology.io/job/Terasology/job/engine/job/PR-5231/5/testReport/org.terasology.engine.integrationenvironment/ExampleTest/Integration_Tests__without_flaky_tests____testClientConnection__/) failed in the last CI run. This test is one of the other flaky tests in `ExampleTest` that I mentioned in the last sync. We probably want to mark at least that one "flaky", too.

jdrueckert committed:
ci: make flaky integration test stage non-blocking

result of flaky test will be displayed in also github, result of failing
normal test is viewed in jenkins.

jdrueckert commented:
![image](https://github.com/MovingBlocks/Terasology/assets/29981695/5ed7e7c7-6043-45b2-a8c7-8ac29cce3791)

The delay manager test fails so rarely that I wouldn't mark that one flaky, but the ExampleTests fail more frequently, so let's mark those flaky. Amusingly the one already marked flaky didn't fail this time :sweat_smile:

soloturn commented:
ok, committed your suggestions and squashed into one. i am understanding correct, we still see the cause of any failed integration test?

BenjaminAmos commented on engine-tests/build.gradle.kts:
```diff
@@ -2,6 +2,13 @@
+// while locally all tests should be run, when building via github, tests can be run separated in the github pipeline
+// file. for integrationtests a tag "flaky" was introduced to mark tests which frequently fail pipelines
+//      gradle test
+//      gradle --consoleplan unitTest
```
```suggestion
// while locally all tests should be run, when building via github, tests can be run separated in the github pipeline
// file. For integration tests, a tag "flaky" was introduced to mark tests which frequently fail pipelines
//      gradle test
//      gradle --console=plain unitTest
```

BenjaminAmos reviewed (dismissed):
There shouldn't be any flaky tests, however, those flaking tests have little benefit and the failures have been very disruptive. This is likely the most practical path forward right now.

soloturn committed:
Update engine-tests/build.gradle.kts

Co-authored-by: BenjaminAmos <24301287+BenjaminAmos@users.noreply.github.com>

jdrueckert commented:
> i am understanding correct, we still see the cause of any failed integration test?

in the sense that test failures are still reported including stack trace etc.? yes
but they shouldn't fail the pipeline anymore.
